File: frontend/js/views/multi_product_breakeven.js
// Multi-product break-even — weighted-average contribution margin CVP across a
// product mix, via /calc/multi-product-breakeven. Live as you type.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';

const money = (n) => (n == null ? '—' : '$' + Number(n).toLocaleString(undefined, { maximumFractionDigits: 2 }));
const units = (n) => (n == null ? '—' : Number(n).toLocaleString(undefined, { maximumFractionDigits: 1 }));
const pct = (n) => (n == null ? '—' : Number(n).toLocaleString(undefined, { maximumFractionDigits: 2 }) + '%');

const SEED = [
    { name: 'A', price: 100, vc: 60, mix: 3 },
    { name: 'B', price: 50, vc: 30, mix: 2 },
];

function rowHtml(p, idx) {
    return `
        <div class="mpb-row" data-idx="${idx}">
            <input type="text" class="mpb-name" placeholder="${t('view.mpbe.ph.name')}" value="${p.name || ''}">
            <input type="number" step="0.01" class="mpb-price" placeholder="${t('view.mpbe.ph.price')}" value="${p.price}">
            <input type="number" step="0.01" class="mpb-vc" placeholder="${t('view.mpbe.ph.vc')}" value="${p.vc}">
            <input type="number" step="0.1" class="mpb-mix" placeholder="${t('view.mpbe.ph.mix')}" value="${p.mix}">
            <button type="button" class="mpb-del" data-i18n="view.mpbe.remove">Remove</button>
        </div>`;
}

export async function renderMultiProductBreakeven(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.mpbe.h1.title">// MULTI-PRODUCT BREAK-EVEN</span></h1>
        <p class="muted small" data-i18n="view.mpbe.hint.intro">
            When a business sells several products in a fixed mix, the break-even point depends on the
            blended margin. Each product contributes price minus variable cost per unit; weighting those
            by the unit sales mix gives the weighted-average contribution margin, and fixed costs over
            that is the total break-even volume. Updates as you type.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.mpbe.h2.inputs">The product mix</h2>
            <form id="mpbe-form" class="inline-form">
                <label><span data-i18n="view.mpbe.label.fixed">Fixed costs ($)</span>
                    <input type="number" step="1000" min="0" name="fixed_costs_usd" value="60000" required></label>
            </form>
            <div class="mpb-head">
                <span data-i18n="view.mpbe.col.name">Product</span>
                <span data-i18n="view.mpbe.col.price">Price</span>
                <span data-i18n="view.mpbe.col.vc">Var. cost</span>
                <span data-i18n="view.mpbe.col.mix">Mix units</span>
                <span></span>
            </div>
            <div id="mpbe-rows">${SEED.map(rowHtml).join('')}</div>
            <button type="button" id="mpbe-add" class="secondary" data-i18n="view.mpbe.add">+ Add product</button>
        </div>
        <div id="mpbe-result" class="lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const rowsEl = mount.querySelector('#mpbe-rows');
    const fixedInput = mount.querySelector('input[name="fixed_costs_usd"]');

    const generate = async () => {
        const products = [...rowsEl.querySelectorAll('.mpb-row')].map((row) => ({
            name: row.querySelector('.mpb-name').value || '',
            price_usd: Number(row.querySelector('.mpb-price').value) || 0,
            variable_cost_usd: Number(row.querySelector('.mpb-vc').value) || 0,
            mix_units: Number(row.querySelector('.mpb-mix').value) || 0,
        })).filter((p) => p.mix_units > 0);
        const body = { fixed_costs_usd: Number(fixedInput.value) || 0, products };
        if (!products.length) {
            mount.querySelector('#mpbe-result').innerHTML = '';
            return;
        }
        try {
            const r = await api.calcMultiProductBreakeven(body);
            if (!viewIsCurrent(tok)) return;
            renderResult(mount, r);
        } catch (err) {
            showToast(err.message || t('view.mpbe.toast.error'), { level: 'error' });
        }
    };
    const debounced = debounce(generate, 250);

    mount.addEventListener('input', debounced);
    mount.querySelector('#mpbe-add').addEventListener('click', () => {
        rowsEl.insertAdjacentHTML('beforeend', rowHtml({ name: '', price: '', vc: '', mix: '' }, rowsEl.children.length));
        applyUiI18n(rowsEl.lastElementChild);
        generate();
    });
    rowsEl.addEventListener('click', (e) => {
        if (e.target.classList.contains('mpb-del')) {
            e.target.closest('.mpb-row').remove();
            generate();
        }
    });
    generate();
}

function renderResult(mount, r) {
    const el = mount.querySelector('#mpbe-result');
    if (!r.is_feasible) {
        el.innerHTML = `<div class="chart-panel"><p class="neg" data-i18n="view.mpbe.infeasible">No break-even — the blended contribution margin is not positive.</p></div>`;
        applyUiI18n(el);
        return;
    }
    const rows = r.products.map((p) => `
        <tr>
            <td>${p.name || '—'}</td>
            <td>${money(p.contribution_margin_usd)}</td>
            <td>${pct(p.mix_proportion_pct)}</td>
            <td>${units(p.breakeven_units)}</td>
            <td>${money(p.breakeven_revenue_usd)}</td>
        </tr>`).join('');
    el.innerHTML = `
        <div class="chart-panel">
            <h2 data-i18n="view.mpbe.h2.result">The break-even</h2>
            <div class="cards">
                <div class="card pos"><div class="label" data-i18n="view.mpbe.card.units">Break-even units</div>
                    <div class="value pos">${units(r.breakeven_units_total)}</div></div>
                <div class="card"><div class="label" data-i18n="view.mpbe.card.revenue">Break-even revenue</div>
                    <div class="value">${money(r.breakeven_revenue_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.mpbe.card.wacm">Weighted-avg CM</div>
                    <div class="value">${money(r.weighted_avg_cm_usd)}</div></div>
            </div>
            <table class="data-table">
                <tbody>
                    <tr><td data-i18n="view.mpbe.row.wacm">Weighted-average CM / unit</td><td>${money(r.weighted_avg_cm_usd)}</td></tr>
                    <tr><td data-i18n="view.mpbe.row.wap">Weighted-average price</td><td>${money(r.weighted_avg_price_usd)}</td></tr>
                    <tr><td data-i18n="view.mpbe.row.ratio">Weighted CM ratio</td><td>${pct(r.weighted_cm_ratio_pct)}</td></tr>
                    <tr class="emph"><td data-i18n="view.mpbe.row.units">Total break-even units</td><td>${units(r.breakeven_units_total)}</td></tr>
                    <tr class="emph"><td data-i18n="view.mpbe.row.revenue">Break-even revenue</td><td>${money(r.breakeven_revenue_usd)}</td></tr>
                </tbody>
            </table>
            <table class="data-table" style="margin-top:10px">
                <thead><tr>
                    <th data-i18n="view.mpbe.th.name">Product</th>
                    <th data-i18n="view.mpbe.th.cm">CM/unit</th>
                    <th data-i18n="view.mpbe.th.mix">Mix</th>
                    <th data-i18n="view.mpbe.th.units">BE units</th>
                    <th data-i18n="view.mpbe.th.revenue">BE revenue</th>
                </tr></thead>
                <tbody>${rows}</tbody>
            </table>
        </div>
    `;
    applyUiI18n(el);
}

File: crates/traderview-core/src/multi_product_breakeven.rs
//! Multi-product break-even — cost-volume-profit analysis when a business sells
//! several products in a fixed sales mix.
//!
//! Each product contributes (price − variable cost) per unit; weighting those by
//! the unit sales mix gives a weighted-average contribution margin per composite
//! unit. Fixed costs over that WACM is the total break-even unit volume, which
//! splits back across the products by their mix proportions.
//!
//! ```text
//! WACM            = Σ (price − variable cost)ᵢ × mix proportionᵢ
//! break-even units = fixed costs / WACM
//! unitsᵢ          = break-even units × mix proportionᵢ
//! ```
//!
//! Distinct from `break-even` (a single product) — the mix is what makes the
//! blended margin, so the answer shifts as the mix shifts.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct Product {
    #[serde(default)]
    pub name: String,
    pub price_usd: f64,
    pub variable_cost_usd: f64,
    /// Relative units of this product in the sales mix.
    pub mix_units: f64,
}

#[derive(Debug, Clone, Deserialize)]
pub struct MultiProductInput {
    pub fixed_costs_usd: f64,
    pub products: Vec<Product>,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct ProductResult {
    pub name: String,
    pub contribution_margin_usd: f64,
    pub mix_proportion_pct: f64,
    pub breakeven_units: Option<f64>,
    pub breakeven_revenue_usd: Option<f64>,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct MultiProductResult {
    /// Weighted-average contribution margin per composite unit.
    pub weighted_avg_cm_usd: f64,
    pub weighted_avg_price_usd: f64,
    /// WACM / weighted-average price.
    pub weighted_cm_ratio_pct: Option<f64>,
    pub breakeven_units_total: Option<f64>,
    pub breakeven_revenue_usd: Option<f64>,
    pub products: Vec<ProductResult>,
    /// True when the blended margin is positive (a break-even exists).
    pub is_feasible: bool,
}

pub fn analyze(input: &MultiProductInput) -> MultiProductResult {
    let total_mix: f64 = input.products.iter().map(|p| p.mix_units).sum();

    let mut wacm = 0.0;
    let mut wap = 0.0;
    if total_mix > 0.0 {
        for p in &input.products {
            let prop = p.mix_units / total_mix;
            wacm += (p.price_usd - p.variable_cost_usd) * prop;
            wap += p.price_usd * prop;
        }
    }

    let feasible = wacm > 0.0;
    let be_units_total = if feasible {
        Some(input.fixed_costs_usd / wacm)
    } else {
        None
    };

    let products = input
        .products
        .iter()
        .map(|p| {
            let prop = if total_mix > 0.0 {
                p.mix_units / total_mix
            } else {
                0.0
            };
            let units = be_units_total.map(|t| t * prop);
            ProductResult {
                name: p.name.clone(),
                contribution_margin_usd: p.price_usd - p.variable_cost_usd,
                mix_proportion_pct: prop * 100.0,
                breakeven_units: units,
                breakeven_revenue_usd: units.map(|u| u * p.price_usd),
            }
        })
        .collect::<Vec<_>>();

    let be_revenue = if be_units_total.is_some() {
        Some(products.iter().filter_map(|p| p.breakeven_revenue_usd).sum())
    } else {
        None
    };

    MultiProductResult {
        weighted_avg_cm_usd: wacm,
        weighted_avg_price_usd: wap,
        weighted_cm_ratio_pct: if wap > 0.0 {
            Some(wacm / wap * 100.0)
        } else {
            None
        },
        breakeven_units_total: be_units_total,
        breakeven_revenue_usd: be_revenue,
        products,
        is_feasible: feasible,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 1e-2
    }

    fn two_products() -> MultiProductInput {
        MultiProductInput {
            fixed_costs_usd: 60_000.0,
            products: vec![
                Product { name: "A".into(), price_usd: 100.0, variable_cost_usd: 60.0, mix_units: 3.0 },
                Product { name: "B".into(), price_usd: 50.0, variable_cost_usd: 30.0, mix_units: 2.0 },
            ],
        }
    }

    #[test]
    fn weighted_avg_cm() {
        // 40 × 0.6 + 20 × 0.4 = 32.
        assert!(close(analyze(&two_products()).weighted_avg_cm_usd, 32.0));
    }

    #[test]
    fn weighted_avg_price_and_ratio() {
        let r = analyze(&two_products());
        // 100 × 0.6 + 50 × 0.4 = 80; ratio 32 / 80 = 40%.
        assert!(close(r.weighted_avg_price_usd, 80.0));
        assert!(close(r.weighted_cm_ratio_pct.unwrap(), 40.0));
    }

    #[test]
    fn breakeven_units_total() {
        // 60,000 / 32 = 1,875.
        assert!(close(analyze(&two_products()).breakeven_units_total.unwrap(), 1_875.0));
    }

    #[test]
    fn breakeven_revenue() {
        // 1,875 units split 60/40 → A 1,125 × 100 + B 750 × 50 = 150,000.
        // Cross-check: 60,000 / 0.40 = 150,000.
        assert!(close(analyze(&two_products()).breakeven_revenue_usd.unwrap(), 150_000.0));
    }

    #[test]
    fn per_product_units() {
        let r = analyze(&two_products());
        assert!(close(r.products[0].breakeven_units.unwrap(), 1_125.0));
        assert!(close(r.products[1].breakeven_units.unwrap(), 750.0));
    }

    #[test]
    fn per_product_proportions() {
        let r = analyze(&two_products());
        assert!(close(r.products[0].mix_proportion_pct, 60.0));
        assert!(close(r.products[1].mix_proportion_pct, 40.0));
    }

    #[test]
    fn per_product_contribution_margin() {
        let r = analyze(&two_products());
        assert!(close(r.products[0].contribution_margin_usd, 40.0));
        assert!(close(r.products[1].contribution_margin_usd, 20.0));
    }

    #[test]
    fn single_product_reduces_to_simple() {
        let r = analyze(&MultiProductInput {
            fixed_costs_usd: 40_000.0,
            products: vec![Product {
                name: "Only".into(),
                price_usd: 100.0,
                variable_cost_usd: 60.0,
                mix_units: 1.0,
            }],
        });
        // 40,000 / 40 = 1,000 units.
        assert!(close(r.breakeven_units_total.unwrap(), 1_000.0));
        assert!(close(r.weighted_cm_ratio_pct.unwrap(), 40.0));
    }

    #[test]
    fn infeasible_when_margin_nonpositive() {
        let r = analyze(&MultiProductInput {
            fixed_costs_usd: 60_000.0,
            products: vec![Product {
                name: "Loss".into(),
                price_usd: 50.0,
                variable_cost_usd: 60.0,
                mix_units: 1.0,
            }],
        });
        assert!(!r.is_feasible);
        assert!(r.breakeven_units_total.is_none());
        assert!(r.breakeven_revenue_usd.is_none());
    }

    #[test]
    fn mix_shift_changes_breakeven() {
        // Heavier weight on the high-margin product lowers break-even units.
        let mut heavy = two_products();
        heavy.products[0].mix_units = 8.0; // more of A (cm 40)
        let base_units = analyze(&two_products()).breakeven_units_total.unwrap();
        let heavy_units = analyze(&heavy).breakeven_units_total.unwrap();
        assert!(heavy_units < base_units);
    }
}
